scipy/stats/tests/test_hypotests.py

```diff
@@ -405,3 +406,150 @@ def test_contingency_table_return(self):
         y = np.arange(10)
         res = stats.somersd(x, y)
         assert_equal(res.table, np.eye(10))
+
+
+class TestBarnardExact(object):
+    """Some tests to show that barnard_exact() works correctly."""
+
+    @pytest.mark.parametrize(
+        "input_sample,expected",
+        [
+            ([[43, 40], [10, 39]], (3.555406779643, 0.000362832367)),
+            ([[100, 2], [1000, 5]], (-1.776382925679, 0.135126970878)),
+            ([[2, 7], [8, 2]], (-2.518474945157, 0.019210815430)),
+            ([[5, 1], [10, 10]], (1.449486150679, 0.156277546306)),
+            ([[5, 15], [20, 20]], (-1.851640199545, 0.066363501421)),
+            ([[5, 16], [20, 25]], (-1.609639949352, 0.116984852192)),
+            ([[10, 5], [10, 1]], (-1.449486150679, 0.177536588915)),
+            ([[5, 0], [1, 4]], (2.581988897472, 0.013671875000)),
+            ([[0, 1], [3, 2]], (-1.095445115010, 0.509667991877)),
+            ([[0, 2], [6, 4]], (-1.549193338483, 0.197019618792)),
+            ([[2, 7], [8, 2]], (-2.518474945157, 0.019210815430)),
+        ],
+    )
+    def test_precise(self, input_sample, expected):
+        """The expected values have been generated by R, using a resolution
+        for the nuisance parameter of 1e-6 :
+        ```R
+        library(Barnard)
+        options(digits=10)
+        barnard.test(43, 40, 10, 39, dp=1e-6, pooled=True)
+        ```
+        """
+        res = barnard_exact(input_sample)
+        statistic, pvalue = res.statistic, res.pvalue
+        assert_allclose([statistic, pvalue], expected)
+
+    @pytest.mark.parametrize(
+        "input_sample,expected",
+        [
+            ([[43, 40], [10, 39]], (3.920362887717, 0.000289470662)),
+            ([[100, 2], [1000, 5]], (-1.139432816087, 0.950272080594)),
+            ([[2, 7], [8, 2]], (-3.079373904042, 0.020172119141)),
+            ([[5, 1], [10, 10]], (1.622375939458, 0.150599922226)),
+            ([[5, 15], [20, 20]], (-1.974771239528, 0.063038448651)),
+            ([[5, 16], [20, 25]], (-1.722122973346, 0.133329494287)),
+            ([[10, 5], [10, 1]], (-1.765469659009, 0.250566655215)),
+            ([[5, 0], [1, 4]], (5.477225575052, 0.007812500000)),
+            ([[0, 1], [3, 2]], (-1.224744871392, 0.509667991877)),
+            ([[0, 2], [6, 4]], (-1.732050807569, 0.197019618792)),
+            ([[2, 7], [8, 2]], (-3.079373904042, 0.020172119141)),
+        ],
+    )
+    def test_pooled_param(self, input_sample, expected):
+        """The expected values have been generated by R, using a resolution
+        for the nuisance parameter of 1e-6 :
```

Comment:
I see. Before reading this, I did not really understand how `n_samples` was used. Now, after reading this and the documentation for [`shgo`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.shgo.html), I have a better understanding.

Do you think the R function is doing some sort of brute force maximization with the given resolution `dp`?

I know it is probably in one of the 300 comments in the PR, but why is `shgo` the preferred optimizer? Is the function not very well behaved w.r.t. `pi` such that a global optimizer is needed? I guess I'm surprised since it's a scalar, and I wouldn't have thought that a global optimizer would be needed unless the function is pretty crazy looking.

Reply:
I suggested to do this. Originally it was a brute force. I tested several methods and shgo provided good results in comparison. So yes the function is quite complicated, lots of peaks and valleys.

There is quite some benchmarks in the discussions about this and how to set the parameters.

Reply:
Thanks! Just curious.

Reply:
Thanks @tupui, I couldn't have said it better